components/cam/src/physics/cam/phys_grid.F90

```diff
@@ -4610,15 +4834,55 @@ subroutine create_chunks(opt, chunks_per_thread)
             maxcol_chk(smp) = 0
             maxcol_chks(smp) = 0
          endif
-      enddo      
+      enddo
+#ifndef PPCOLS
+!
+! If column cost is provided and pcols is not specified (and opt /= 4),
+! then can potentially get better load balanced chunks by picking 
+! maxcol_chk, maxcol_chks, and pcols_vsmp as large as possible before
+! chunk creation. After the chunks are created, can then set pcols_vsmp
+! to the actual sizes generated (per virtual smp). There needs to be
+! a bound on this initial max pcols value to guard against memory
+! issues. Use cols_per_chnk_max if this is specified. Otherwise just
+! stick with the previously calculated maxcol_chk, maxcol_chks, and
+! pcols_vsmp and skip the following logic.
+! Note that already used the previously calculated pcols_vsmp to determine
+! the number of chunks, so this optimization is still constrained by this.
+!
+      if (cols_per_chnk <= 0) then
+         if ((use_cost_d) .and. (cols_per_chnk_max > 0) .and. (opt /= 4)) then
+            max_pcols = cols_per_chnk_max
+! decrease (if necessary) to be a multiple of cols_per_chnk_mult
+            if ((cols_per_chnk_mult > 1) .and. (cols_per_chnk_mult <= cols_per_chnk_max)) then
+               max_pcols = &
+                  cols_per_chnk_mult*floor(real(max_pcols,r8)/real(cols_per_chnk_mult,r8))
+            endif
+! If the new max_pcols is larger than the previously calculated pcols_vsmp
+! for a given virtual smp, then set pcols_vsml(smp) to max_pcols.
+            do smp=0,nvsmp-1
+               if (max_pcols > pcols_vsmp(smp)) then
+                  pcols_vsmp(smp) = max_pcols
+               endif
+            enddo
+! Reset maxcol_chk(smp) to the updated pcols_vsmp(smp), and reset maxcol_chks(smp)
+! to be the number of processes assigned to the virtual smp
+            maxcol_chk(:) = pcols_vsmp(:)
+            maxcol_chks(:) = nvsmpchunks(:)
+! Reset max_pcols. Note this could be larger than necessary (after resetting
+! pcols_vsmp based on the actual number of columns assigned to each chunk), 
+! but it is needed now in order to allocate the gcol array in the
+! chunks data structure. The gcol array will be reallocated after the return
+! from the create_chunks routine, after which there will be no wasted space.
+            max_pcols = maxval(pcols_vsmp)
+         endif
+      endif
+#endif
 !
 ! Allocate chunks and knuhcs data structures
```

Comment:
I realize this is perpendicular to this PR, but what's the point of the 'knuhcs' variable? I've never noticed it, but I imagine it could probably have a more accurately descriptive name.

Reply:
inverse of chunks. This is a very old choice (mine, trying to be clever I guess).